### ISSUE TITLE
Reaper upgrade can cause schema migration exception

### ISSUE THREAD

**Issue by jsanda** [k8ssandra/reaper-operator#63]
**Describe the bug**
I observed the issue in an a k8ssandra integration test for upgrades. It upgrades from k8ssandra 1.0 to the latest which right now is 1.3.0. This involves a Reaper upgrade. While investigating some test failures I saw this exception in the Reaper log:

```
ERROR  [2021-08-17 20:48:34,020] [main] i.c.ReaperApplication - Storage is not ready yet, trying again to connect shortly...
org.cognitor.cassandra.migration.MigrationException: Error during migration of script 022_cluster_states.cql while executing 'ALTER TABLE cluster ADD state text;'
        at org.cognitor.cassandra.migration.Database.execute(Database.java:269)
        at java.util.Collections$SingletonList.forEach(Collections.java:4824)
        at org.cognitor.cassandra.migration.MigrationTask.migrate(MigrationTask.java:68)
        at io.cassandrareaper.storage.CassandraStorage.migrate(CassandraStorage.java:366)
        at io.cassandrareaper.storage.CassandraStorage.initializeCassandraSchema(CassandraStorage.java:297)
        at io.cassandrareaper.storage.CassandraStorage.initializeAndUpgradeSchema(CassandraStorage.java:255)
        at io.cassandrareaper.storage.CassandraStorage.<init>(CassandraStorage.java:243)
        at io.cassandrareaper.storage.InitializeStorage.initializeStorageBackend(InitializeStorage.java:69)
        at io.cassandrareaper.ReaperApplication.tryInitializeStorage(ReaperApplication.java:486)
        at io.cassandrareaper.ReaperApplication.run(ReaperApplication.java:173)
        at io.cassandrareaper.ReaperApplication.run(ReaperApplication.java:90)
        at io.dropwizard.cli.EnvironmentCommand.run(EnvironmentCommand.java:43)
        at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:87)
        at io.dropwizard.cli.Cli.run(Cli.java:78)
        at io.dropwizard.Application.run(Application.java:93)
        at io.cassandrareaper.ReaperApplication.main(ReaperApplication.java:109)
Caused by: com.datastax.driver.core.exceptions.InvalidQueryException: Invalid column name state because it conflicts with an existing column
        at com.datastax.driver.core.exceptions.InvalidQueryException.copy(InvalidQueryException.java:50)
```

I also observed that there were two Reaper pods. The reason that there were two Reaper pods is because the default Deployment update strategy is used. It looks like this:

```yaml
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
```

Here is what the Deployment [docs](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) say about updates:

>Deployment ensures that only a certain number of Pods are down while they are being updated. By default, it ensures that at least 75% of the desired number of Pods are up (25% max unavailable).

We should be able to resolve this by changing the `DeploymentStrategyType` to `Recreate`. This will kill existing pods before creating any new ones.

**To Reproduce**

1. Deploy k8ssandra 1.0
1. Upgrade to k8ssandra 1.3
1. Monitor the Reaper deployment
1. Check the Reaper logs for the above error

I am seeing this error frequently in GitHub Actions but not so much locally.

**Expected behavior**
There should only be a single Reaper pod running at any given time.

**Environment (please complete the following information):**

* reaper-operator version:
0.3.4

* Helm charts version info 
1.3.0

┆Issue is synchronized with this [Jira Bug](https://k8ssandra.atlassian.net/browse/K8SSAND-816) by [Unito](https://www.unito.io)
┆Affected Versions: k8ssandra-1.3.0
┆Fix Versions: k8ssandra-1.4.0
┆Issue Number: K8SSAND-816
┆Priority: Medium

**Comment by Miles-Garnsey** [comment]
From discussions with @jsanda today (thank you again for walking me through some of this!) 

We think that at [line 264](https://github.com/k8ssandra/reaper-operator/blob/a129a522a4ed9c1e3e8fdcfb9ee6728807cd20cf/pkg/reconcile/reconcilers.go#L264) of the ReconcileDeployment function, there should be a check for the deployment strategy type in the existing deployment. 

If the strategy type is not RecreateDeploymentStrategyType then we should use `client.Patch()` to update it to RecreateDeploymentStrategyType before immediately returning from the reconciler and re-queueing the request.

I'm still getting my head around the APIs, but we think we want something of the form - 

```
if deployment.Spec.Strategy.Type != appsv1.RecreateDeploymentStrategyType {
		req.Logger.Error(err, "Reaper deployment not currently using appsv1.RecreateDeploymentStrategyType, patching and re-queueing the request")
		patch := client.MergeFrom(deployment.DeepCopy())
		deployment.Spec.Strategy = appsv1.DeploymentStrategy{
			Type: appsv1.RecreateDeploymentStrategyType,
		}
		if err := r.Patch(ctx, deployment, patch); err != nil {
			req.Logger.Error(err, "Failed to patch deployment with appsv1.RecreateDeploymentStrategyType")
		}
		return &ctrl.Result{RequeueAfter: 10 * time.Second}, err
	}
```

Now I'm off to test this and see if it works.

**Comment by Miles-Garnsey** [comment]
Patch is manually tested and [PR raised](https://github.com/k8ssandra/reaper-operator/pull/65). @jdonenine depending on how urgent this is I can put the tests in a different PR if you want to get this one merged right away? 

My testing has focused on basically just changing the deployment strategy to rolling and then watching the controller change it back, while also confirming that no other changes are made to the deployment until the strategy is changed. I think we can automate a test that simply does that.

I'm now going to do a manual test of the upgrade process to see if it breaks, but I know this behaviour is inconsistent so we'll have to see if I can observe the problem.

**Comment by Miles-Garnsey** [comment]
After testing this quite extensively yesterday I've come to the conclusion that there is more than one error here. I was able to determine this by using the new Docker virtualisation framework, which is so slow that it replicates the resource constraints we see on GHA.

# Error 1: timeouts during migration deadlock Reaper

On K8ssandra 1.0.0 - 1.2.0 Reaper is frequently restarted during the initial migration. This occurs because during migration it does not respond to liveness/health probes, k8s marks it as down and it restarts its container. 

But Reaper acquires a lock to complete the migration, and doesn't release it when shut down. Consequently, the subsequent instance that is spun up cannot make progress on the migration and we see errors of the form `INFO   [2021-08-23 07:28:04,844] [main] o.c.c.m.Database - Schema migration is locked by another instance. Waiting for it to be released...`. Eventually we see the container restarted because it will not respond to probes.

This problem is fixed by #57 which performs the migration in a non health-checked/liveness-probed initContainer. But the consequence is that in resource constrained environments, any e2e upgrade test from 1.1.0 or 1.2.0 will non-deterministically fail. This is not because of the upgrade, it is because the original Reaper deployment could never attain a healthy state.

# Error 2: Multiple pods on upgrade due to the use of a rolling deployment

Because Reaper is lock heavy in other ways, having multiple instances of the pods also appears to cause other deadlocks as reported by @jsanda above. 

*However, I think the error we're seeing here actually refers to Cassandra not being available, rather than any sort of contention issues. @jsanda can you confirm how many nodes you were running when you saw the below?*

```
ERROR  [2021-08-17 20:48:34,020] [main] i.c.ReaperApplication - Storage is not ready yet, trying again to connect shortly...
org.cognitor.cassandra.migration.MigrationException: Error during migration of script 022_cluster_states.cql while executing 'ALTER TABLE cluster ADD state text;'
```

If it was a single node, I can reproduce this, and it is occurring because Cassandra is 100% down during the upgrade. I'm a bit confused as to why that is, because I'd have thought that Cassandra would be under a rollingRestart deployment strategy and therefore during a rollout the old pod should be kept around. We need to look into why that is not the case.

# A new patch

The patch I've assembled under PR #65 hopes to resolve any potential contention that may be caused by multiple Reaper instances running. 

I've confirmed that it does the expected things both at upgrade (where we see it switch the deployment strategy to recreate, re-queue and then apply the remaining changes) and in the case of manual changes to the deploymentStrategy.

We get this message when the new code path is triggered - 
```
2021-08-24T04:37:29.405Z	INFO	controllers.Reaper	Reaper deploymentk8ssandra-0-reaper not currently using appsv1.RecreateDeploymentStrategyType, patching and re-queueing the request{"reaper": "default/k8ssandra-0-reaper"}
```

I do want to mention that in the case of error (1) above it will not help as it is a patch on 3.1.0. In the case of (2) above, I suspect that we need to look at why Cassandra is going down without following the rolling update strategy we'd expect.

**Comment by Miles-Garnsey** [comment]
Confirming I do not see this error when running a 3 node cluster.

**Comment by Miles-Garnsey** [comment]
Also, I keep looking at the docs and for some reason the behaviour from the Deployment really doesn't match the doco for the Recreate deployment strategy. Running kubectl get pods every 2 seconds I see the below - 

```
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-6bd5d66557-htd4v             1/1     Running       0          78m
k8ssandra-0-cass-operator-998c64dc6-jv7b2              0/1     Running       0          4s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          8s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          2/2     Running       0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          3s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-admission-patch-qmw94       1/1     Running       0          1s
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          2s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-57f48d4cd5-c4xgx           0/1     Terminating   0          78m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          4s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
NAME                                                   READY   STATUS      RESTARTS   AGE
k8ssandra-0-cass-operator-6bd5d66557-htd4v             1/1     Running     0          78m
k8ssandra-0-cass-operator-998c64dc6-jv7b2              0/1     Running     0          6s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed   0          10s
k8ssandra-0-dc1-default-sts-0                          2/2     Running     0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running     0          78m
k8ssandra-0-dc1-default-sts-2                          2/2     Running     0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running     0          5s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running     7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running     0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running     0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running     0          4s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running     0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running     0          6s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running     0          78m
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-6bd5d66557-htd4v             1/1     Terminating   0          78m
k8ssandra-0-cass-operator-998c64dc6-jv7b2              1/1     Running       0          8s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          12s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          2/2     Terminating   0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          7s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          6s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          8s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-6bd5d66557-htd4v             0/1     Terminating   0          78m
k8ssandra-0-cass-operator-998c64dc6-jv7b2              1/1     Running       0          10s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          14s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          2/2     Terminating   0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          9s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          8s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          10s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-6bd5d66557-htd4v             0/1     Terminating   0          78m
k8ssandra-0-cass-operator-998c64dc6-jv7b2              1/1     Running       0          13s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          17s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          2/2     Terminating   0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          12s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          11s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          13s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-6bd5d66557-htd4v             0/1     Terminating   0          78m
k8ssandra-0-cass-operator-998c64dc6-jv7b2              1/1     Running       0          15s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          19s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          2/2     Terminating   0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          14s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          13s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          15s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-6bd5d66557-htd4v             0/1     Terminating   0          78m
k8ssandra-0-cass-operator-998c64dc6-jv7b2              1/1     Running       0          18s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          22s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          2/2     Terminating   0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          17s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          16s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          18s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-998c64dc6-jv7b2              1/1     Running       0          20s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          24s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          2/2     Terminating   0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          19s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          18s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          20s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-998c64dc6-jv7b2              1/1     Running       0          23s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          27s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          2/2     Terminating   0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          22s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          21s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          23s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-998c64dc6-jv7b2              1/1     Running       0          25s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          29s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          2/2     Terminating   0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          24s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          23s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          25s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-998c64dc6-jv7b2              1/1     Running       0          27s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          31s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          2/2     Terminating   0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          26s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          25s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          27s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-998c64dc6-jv7b2              1/1     Running       0          30s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          34s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          2/2     Terminating   0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          29s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          28s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          30s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-998c64dc6-jv7b2              1/1     Running       0          33s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          37s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          1/2     Terminating   0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          32s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          31s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          33s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-998c64dc6-jv7b2              1/1     Running       0          35s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          39s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          1/2     Terminating   0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          34s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          33s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          35s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
NAME                                                   READY   STATUS        RESTARTS   AGE
k8ssandra-0-cass-operator-998c64dc6-jv7b2              1/1     Running       0          37s
k8ssandra-0-crd-upgrader-job-k8ssandra-8dhww           0/1     Completed     0          41s
k8ssandra-0-dc1-default-sts-0                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-1                          2/2     Running       0          78m
k8ssandra-0-dc1-default-sts-2                          1/2     Terminating   0          78m
k8ssandra-0-dc1-stargate-5575d468c-nk4gp               0/1     Running       0          36s
k8ssandra-0-dc1-stargate-66c4cf96b8-zcvmm              1/1     Running       7          78m
k8ssandra-0-grafana-749d4b96c-pcvqc                    2/2     Running       0          78m
k8ssandra-0-kube-prometheu-operator-694b8877bc-d46st   1/1     Running       0          78m
k8ssandra-0-reaper-56d5f6c786-f4ngq                    0/1     Running       0          35s
k8ssandra-0-reaper-99f768896-kgnzf                     1/1     Running       0          72m
k8ssandra-0-reaper-operator-7d7d95f948-pv76g           1/1     Running       0          37s
prometheus-k8ssandra-0-kube-prometheu-prometheus-0     2/2     Running       0          78m
```

**Comment by Miles-Garnsey** [comment]
To explain the above: the k8s documentation is a bit misleading when it says that the recreate strategy will ensure that the old pod is completely removed prior to spinning up the new pod. 

The `DeploymentController` will call [syncDeployment](https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/controller/deployment/deployment_controller.go#L652) which in turn calls [rolloutRecreate](https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/controller/deployment/recreate.go#L47). We can see that rolloutRecreate ends up calling [oldPodsRunning](https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/pkg/controller/deployment/recreate.go#L98) to determine whether to create a new replicaSet. 

But `oldPodsRunning` doesn't actually count a pod that is in a terminal state as running, so there is at least one condition where two replicaSets may exist concurrently.

We tested whether this was the reason for two pods being created, but it doesn't appear to be. We can see from the below that the incoming pod is initialised at `06:54:13` while the outgoing container is terminated in an error state at `06:54:56`. 

While both containers were never running concurrently, I am still looking for evidence that this is impossible.

Outgoing pod
```
Wed 25 Aug 2021 16:55:06 AEST
conditions:
  - lastProbeTime: null
    lastTransitionTime: "2021-08-25T06:05:07Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2021-08-25T06:54:57Z"
    message: 'containers with unready status: [reaper]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2021-08-25T06:54:57Z"
    message: 'containers with unready status: [reaper]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2021-08-25T06:05:07Z"
    status: "True"
    type: PodScheduled
containerStatuses:
  - containerID: containerd://24c38878f03448f378eb0b02883f2abd88f7bf46d4b74012b285dc4bfdc6fab9
    image: docker.io/thelastpickle/cassandra-reaper:2.2.2
    imageID: docker.io/thelastpickle/cassandra-reaper@sha256:752f041e7c933602b052e8c319aef8c9770cd8073a739b7d35233cd230c3eb62
    lastState:
      terminated:
        containerID: containerd://db018061b16c092537ba03296acb64fffa50bc66603d317d3887da35da97094a
        exitCode: 143
        finishedAt: "2021-08-25T06:54:56Z"
        reason: Error
        startedAt: "2021-08-25T06:17:14Z"
    name: reaper
    ready: false
    restartCount: 5
    started: true
    state:
      running:
        startedAt: "2021-08-25T06:54:57Z"
```

We can then compare against the incoming pod -

```
conditions:
  - lastProbeTime: null
    lastTransitionTime: "2021-08-25T06:54:13Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2021-08-25T06:54:13Z"
    message: 'containers with unready status: [reaper]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2021-08-25T06:54:13Z"
    message: 'containers with unready status: [reaper]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2021-08-25T06:54:13Z"
    status: "True"
    type: PodScheduled
containerStatuses:
  - image: docker.io/thelastpickle/cassandra-reaper:2.3.1
    imageID: ""
    lastState: {}
    name: reaper
    ready: false
    restartCount: 0
    started: false
    state:
      waiting:
        reason: ContainerCreating
```

**Comment by Miles-Garnsey** [comment]
From discussion with Alex, there is no risk that the migration will deadlock for too long due to restarts alone. 

The TTL on the lock table (schema_migration_leader) is set at 5 minutes. As a result, while we may observe `INFO [2021-08-23 07:28:04,844] [main] o.c.c.m.Database - Schema migration is locked by another instance. Waiting for it to be released....` messages for a short time after initContainer failure, this will not be ongoing for very long before the locks are released for a new migration instance to progress the bootstrap.

However, there may still be a case where the migration has commenced but the lock has not been acquired, where we may see the error above - `Caused by: com.datastax.driver.core.exceptions.InvalidQueryException: Invalid column name state because it conflicts with an existing column` we may need to investigate this further if we can make some progress on reproducing it.

**Comment by Miles-Garnsey** [comment]
Note this issue should be fixed by #70 currently awaiting merge.
